File: arm.64/langs/fsharp/src/bind.fs
namespace Pat

open System
open System.IO
open System.Runtime.InteropServices

module Binder =
    let bind (binaryPath: string) : Cabi option =
        let filename = Path.GetFileName binaryPath
        let skips = [ "c"; "cpp"; "rust"; "go"; "swift"; "haskell"; "zig"; "v"; "slint_sample" ]
        if List.contains filename skips then None
        else
            let mutable handle = IntPtr.Zero
            if NativeLibrary.TryLoad(binaryPath, &handle) then
                let mutable pumpPtr = IntPtr.Zero
                if NativeLibrary.TryGetExport(handle, "Pump", &pumpPtr) then
                    let pump = Marshal.GetDelegateForFunctionPointer<PumpDelegate>(pumpPtr)
                    Some { Name = filename; Path = binaryPath; Handle = handle; Pump = pump }
                else
                    NativeLibrary.Free handle
                    None
            else None

File: arm.64/langs/csharp/src/bind.cs
using System.Runtime.InteropServices;

namespace Pat;

public static class Binder
{
    public static Cabi? Bind(string binaryPath)
    {
        string filename = Path.GetFileName(binaryPath);
        string[] skips = { "c", "cpp", "rust", "go", "swift", "haskell", "zig", "v", "slint_sample" };
        if (skips.Contains(filename)) return null;

        if (!NativeLibrary.TryLoad(binaryPath, out IntPtr handle)) return null;
        if (!NativeLibrary.TryGetExport(handle, "Pump", out IntPtr pumpPtr))
        {
            NativeLibrary.Free(handle);
            return null;
        }

        var pump = Marshal.GetDelegateForFunctionPointer<PumpDelegate>(pumpPtr);
        NativeLibrary.TryGetExport(handle, "More", out IntPtr morePtr);
        NativeLibrary.TryGetExport(handle, "Less", out IntPtr lessPtr);

        return new Cabi
        {
            Name = filename,
            Path = binaryPath,
            Handle = handle,
            More = morePtr,
            Pump = pump,
            Less = lessPtr
        };
    }
}
